File: CppGen/CppGen/DataType.cs
using System;
using System.Collections.Generic;

namespace CppGen
{
	// Stores the possible data types of a variable, function argument or return type.
	public class DataType
	{
		public List<Assignment> Assignments = new List<Assignment>();
		public CppType cppType = CppType.VarType;

		// Can be set for debugging purposes the geneate all types as VarType.
		static public bool AllVarType = false;

		// Set automatically when generating code that needs types.
		static public bool IgnoreAllVarType = false;

		// Stores a possible data type and what function it got assigned first.
		public class Assignment
		{
			public Type RawType;
			public string RefId = ""; // Object name
			public DataType ContainerType = null;
			public Function Func;
			public int Line;

			public Assignment(Assignment other, Function func = null, int line = 0)
			{
				RawType = other.RawType;
				RefId = other.RefId;
				if (other.ContainerType != null)
					ContainerType = new DataType(other.ContainerType);
				Func = func;
				Line = line;
			}

			public Assignment(Type type, string refId, DataType containerType, Function func = null, int line = 0)
			{
				RawType = type;
				RefId = refId;
				ContainerType = containerType;
				Func = func;
				Line = line;
			}

			// Returns the assignment in string form with <> for containers/references.
			public override string ToString()
			{
				if (RawType >= Type.Array) // Container
					return RawType.ToString() + "<" + ContainerType.ToString() + ">";
				else if (RawType == Type.Reference) // Reference
					return RawType.ToString() + "<" + RefId + ">";
				else
					return RawType.ToString();
			}
		}

		public enum Type
		{
			Unknown,
			Variant,
			Null,
			Void,
			Bool,
			IntOrReal,
			Integer,
			Real,
			String,
			Reference,
			Array,
			Vector,
			Matrix,
			List,
			AnyMap,
			IntMap,
			StringMap,
			Map,
			Grid,
			Stack,
			Priority
		}

		public enum CppType
		{
			Void,
			RealType,
			IntType,
			BoolType,
			StringType,
			VecType,
			MatrixType,
			ArrType,
			VarType
		}

		// Copy a DataType
		public DataType(DataType other)
		{
			foreach (Assignment ass in other.Assignments)
				Assignments.Add(new Assignment(ass, ass.Func, ass.Line));
			UpdateCppType();
		}

		// Create a generic type
		public DataType(Type rawType = Type.Unknown)
		{
			if (rawType != Type.Unknown)
				Assignments.Add(new Assignment(rawType, "", null));
			UpdateCppType();
		}

		// Create a reference type
		public DataType(Type rawType, string refId)
		{
			if (rawType != Type.Unknown)
				Assignments.Add(new Assignment(rawType, refId, null));
			UpdateCppType();
		}

		// Create a container type
		public DataType(Type rawType, DataType containerType)
		{
			if (rawType != Type.Unknown)
				Assignments.Add(new Assignment(rawType, "", containerType));
			UpdateCppType();
		}

		// Create a DataType from a text string
		public DataType(string name)
		{
			string[] split = name.Replace(">", "").Split('<');
			Type rawType = Type.Unknown;

			switch (split[0])
			{
				case "unknown": rawType = Type.Unknown; break;
				case "VarType":
				case "VarType&":
				case "const VarType&":
				case "variant": rawType = Type.Variant; break;
				case "null": rawType = Type.Null; break;
				case "void": rawType = Type.Void; break;
				case "BoolType":
				case "bool": rawType = Type.Bool; break;
				case "intorreal": rawType = Type.IntOrReal; break;
				case "IntType":
				case "int": rawType = Type.Integer; break;
				case "RealType":
				case "real": rawType = Type.Real; break;
				case "StringType":
				case "string": rawType = Type.String; break;
				case "reference":
				{
					if (split.Length != 2)
					{
						Console.WriteLine("FATAL ERROR: Missing ID for reference.");
						Environment.Exit(1);
					}
					Assignments.Add(new Assignment(Type.Reference, split[1], null));
					UpdateCppType();
					return;
				}
				case "VecType":
				case "vec": rawType = Type.Vector; break;
				case "MatrixType":
				case "matrix": rawType = Type.Matrix; break;
				case "ArrType":
				case "array": rawType = Type.Array; break;
				case "list": rawType = Type.List; break;
				case "anymap": rawType = Type.AnyMap; break;
				case "intmap": rawType = Type.IntMap; break;
				case "stringmap": rawType = Type.StringMap; break;
				case "map": rawType = Type.Map; break;
				case "grid": rawType = Type.Grid; break;
				case "stack": rawType = Type.Stack;  break;
				case "priority": rawType = Type.Priority; break;
				default:
				{
					Console.WriteLine("FATAL ERROR: Unknown data type: {0}", split[0]);
					Environment.Exit(1);
					break;
				}
			}

			if (rawType >= Type.Array) // Container
			{
				DataType containerType;
				if (rawType != Type.Array && IsRawTypeArray(rawType)) // vec or matrix
					containerType = new DataType(Type.Real);

				else if (split.Length > 1)
				{
					string stringEnd = "";
					for (int i = 1; i < split.Length; i++)
						stringEnd += split[i] + (i < split.Length - 1 ? "<" : "");
					containerType = new DataType(stringEnd);
				}
				else
					containerType = new DataType(Type.Unknown);

				Assignments.Add(new Assignment(rawType, "", containerType));
			}
			else if (rawType != Type.Unknown)
				Assignments.Add(new Assignment(rawType, "", null));

			UpdateCppType();
		}

		// Update the C++ type.
		public void UpdateCppType()
		{
			// Update C++ type
			bool isSet = false;
			foreach (Assignment ass in Assignments)
			{
				CppType assType = CppType.VarType;
				switch (ass.RawType)
				{
					case Type.Variant: assType = CppType.VarType; break;
					case Type.Null: assType = CppType.IntType; break;
					case Type.Void: assType = CppType.Void; break;
					case Type.Bool: assType = CppType.BoolType; break;
					case Type.Integer: assType = CppType.IntType; break;
					case Type.Real:
					case Type.IntOrReal: assType = CppType.RealType; break;
					case Type.String: assType = CppType.StringType; break;
					case Type.Reference: assType = CppType.IntType; break; // TODO pointers?
					case Type.Array: assType = CppType.ArrType; break;
					case Type.Vector: assType = CppType.VecType; break;
					case Type.Matrix: assType = CppType.MatrixType; break;
					case Type.List:
					case Type.AnyMap:
					case Type.IntMap:
					case Type.StringMap:
					case Type.Map:
					case Type.Grid:
					case Type.Stack:
					case Type.Priority: assType = CppType.IntType; break;
				}

				if (!isSet)
				{
					cppType = assType;
					isSet = true;
				}
				else if (cppType != assType)
				{
					cppType = CppType.VarType;
					return;
				}
			}
		}

		// Returns the C++ struct of the type.
		public string ToCpp()
		{
			if (AllVarType && !IgnoreAllVarType && cppType != CppType.Void)
				return "VarType";

			switch (cppType)
			{
				case CppType.Void: return "void";
				case CppType.RealType: return "RealType";
				case CppType.IntType: return "IntType";
				case CppType.BoolType: return "BoolType";
				case CppType.StringType: return "StringType";
				case CppType.VecType: return "VecType";
				case CppType.MatrixType: return "MatrixType";
				case CppType.ArrType: return "ArrType";
				case CppType.VarType: return "VarType";
			}
			return "";
		}


		// Converts the type to a C++ member macro.
		public string ToCppMemberMacro()
		{
			switch (cppType)
			{
				case CppType.RealType: return "Real";
				case CppType.IntType: return "Int";
				case CppType.BoolType: return "Bool";
				case CppType.StringType: return "Str";
				case CppType.VecType: return "Vec";
				case CppType.MatrixType: return "Mat";
				case CppType.ArrType: return "Arr";
				case CppType.VarType: return "Var";
			}
			return "";
		}

		// Converts the type to a default value in C++.
		public string ToCppDefaultValue()
		{
			switch (cppType)
			{
				case CppType.RealType: return "0.0";
				case CppType.IntType: return "IntType(0)";
				case CppType.BoolType: return "false";
				case CppType.StringType: return "\"\"";
				case CppType.VecType: return "VecType()";
				case CppType.MatrixType: return "MatrixType()";
				case CppType.ArrType: return "ArrType()";
				case CppType.VarType: return "VarType()";
			}
			return "";
		}

		// Converts the type to the Type character in C++.
		public string ToCppEnum()
		{
			switch (cppType)
			{
				case CppType.RealType: return "REAL_t";
				case CppType.IntType: return "INTEGER_t";
				case CppType.BoolType: return "BOOLEAN_t";
				case CppType.StringType: return "STRING_t";
				case CppType.VecType: return "VECTOR_t";
				case CppType.MatrixType: return "MATRIX_t";
				case CppType.ArrType: return "ARRAY_t";
				case CppType.VarType: return "VARIANT_t";
			}
			return "";
		}

		// Returns whether the type is a C++ VarType.
		public bool IsCppVarType()
		{
			return cppType == CppType.VarType;
		}

		// Returns a string with all type assignments in the project, if more than one.
		public string GetAssignmentsString(string tabs = "")
		{
			string varStr = "";
			foreach (Assignment ass in Assignments)
			{
				varStr += tabs + ass.ToString();
				if (ass.Func != null)
					varStr += " in " + ass.Func.Name + ":" + ass.Line;
				varStr += "\n";
			}
			return varStr;
		}

		// Returns the type in string form with <> for containers/references or "Variant" for multiple types.
		public override string ToString()
		{
			if (Assignments.Count == 0)
				return "Unknown";

			string str = "";
			for (int i = 0; i < Assignments.Count; i++)
				str += (i > 0 ? ", " : "") + Assignments[i].ToString();

			return str;
		}

		// Returns whether the type has not been assigned any raw type.
		public bool IsUnknown()
		{
			return (Assignments.Count == 0);
		}

		// Returns whether the given raw type is a real number.
		static public bool IsRawTypeReal(Type type)
		{
			return (type == Type.IntOrReal ||
					type == Type.Real ||
					type == Type.Integer ||
					type == Type.Bool);
		}

		// Returns whether the given raw type is accessed using [i] or [@i].
		static public bool IsRawTypeArray(Type type)
		{
			return (type == Type.Array || 
					type == Type.Vector ||
					type == Type.Matrix);
		}

		// Returns whether the given raw type is a map.
		static public bool IsRawTypeMap(Type type)
		{
			return (type == Type.AnyMap || 
					type == Type.Map ||
					type == Type.IntMap ||
					type == Type.StringMap);
		}

		// Returns whether the type has been assigned a real number.
		public bool IsReal()
		{
			if (Assignments.Count == 0)
				return false;

			foreach (Assignment ass in Assignments)
				if (IsRawTypeReal(ass.RawType))
					return true;

			return false;
		}

		// Returns whether the type can store other types and use the [] operator.
		public bool IsContainer()
		{
			foreach (Assignment ass in Assignments)
				if (ass.RawType >= Type.Array)
					return true;
			
			return false;
		}

		// Returns the first assignment that matches the raw type, or null if not found.
		public Assignment GetFirstAssignment(Type rawType = Type.Unknown)
		{
			foreach (Assignment ass in Assignments)
				if (rawType == Type.Unknown ||
					ass.RawType == rawType ||
					(rawType == Type.AnyMap && IsRawTypeMap(ass.RawType)))
					return ass;

			return null;
		}

		// Returns all assignments that matches the raw type.
		public List<Assignment> GetAssignments(Type rawType = Type.Unknown)
		{
			List<Assignment> result = new List<Assignment>();
			foreach (Assignment ass in Assignments)
				if (rawType == Type.Unknown || ass.RawType == rawType)
					result.Add(ass);

			return result;
		}

		// Returns an unique object reference id assigned to the type, or empty string if not found.
		public string GetUniqueReferenceId()
		{
			string refId = "";
			bool isSet = false;
			foreach (Assignment ass in Assignments)
			{
				if (ass.RawType == Type.Reference)
				{
					if (isSet) // Multiple assignments found containing references, exit "any"
						return "any";
					
					refId = ass.RefId;
					isSet = true;
				}
			}

			return refId;
		}

		// Returns the map type of the datatype.
		public Type GetMapType()
		{
			foreach (Assignment ass in Assignments)
				if (IsRawTypeMap(ass.RawType))
					return ass.RawType;

			return Type.Unknown;
		}

		// Assign another DataType class to this one. Returns whether the type was updated.
		public bool Assign(DataType inputType, Function func, int line, int containerLevel = 0)
		{
			const int maxContainerLevel = 1;

			if (inputType == null || inputType.Assignments.Count == 0) // Unknown input
				return false;

			bool changed = false;
			foreach (Assignment inputAss in inputType.Assignments) // Iterate input assignments
			{
				bool addNew = true;
				foreach (Assignment ass in Assignments) // Combine/Add assignments
				{
					Type inputRawType = inputAss.RawType;
					if ((IsRawTypeReal(ass.RawType) && IsRawTypeReal(inputRawType)) || // Both real
						(IsRawTypeArray(ass.RawType) && IsRawTypeArray(inputRawType))) // Both array
					{
						addNew = false;
						if (inputRawType > ass.RawType) // Overwrite if larger
						{
							ass.RawType = inputRawType;
							changed = true;
							ass.Func = func;
							ass.Line = line;
						}
					}
					else if (inputRawType == Type.Reference && ass.RawType == Type.Reference) // Both reference
					{
						if (ass.RefId == "" && inputAss.RefId != "") // Unknown reference
						{
							ass.RefId = inputAss.RefId;
							changed = true;
							ass.Func = func;
							ass.Line = line;
							addNew = false; 
						}
						else if (inputAss.RefId == ass.RefId) // Empty input/Same object
							addNew = false;
					}
					else if (IsRawTypeMap(ass.RawType) && IsRawTypeMap(inputRawType) && inputRawType != ass.RawType) // Both map
					{
						addNew = false;
						if (inputRawType > ass.RawType) // Overwrite if larger
						{
							if (ass.RawType == Type.IntMap) // StringMap + IntMap -> Map
								ass.RawType = Type.Map;
							else
								ass.RawType = inputRawType;
							changed = true;
							ass.Func = func;
							ass.Line = line;
						}
					}
					else if (ass.RawType == inputRawType) // Type match
						addNew = false;

					// Assign container
					if ((ass.RawType == Type.Array || ass.RawType >= Type.List) &&
						inputAss.ContainerType != null && containerLevel < maxContainerLevel &&
						ass.ContainerType.Assign(inputAss.ContainerType, func, line, containerLevel + 1))
					{
						changed = true;
						ass.Func = func;
						ass.Line = line;
					}

					if (!addNew)
						break;
				}

				// Other assignment is unique, add a copy to this type
				if (addNew)
				{
					Assignments.Add(new Assignment(inputAss, func, line));
					changed = true;
				}
			}

			if (changed)
				UpdateCppType();

			return changed;
		}
	}
}

File: CppGen/Sources/DataType.cpp
#include "CppGen.hpp"

namespace CppGen
{
DataType::CppType cppTypeFor(DataType::Type rawType)
{
	switch (rawType)
	{
		case DataType::Type::Void: return DataType::CppType::Void;
		case DataType::Type::Real:
		case DataType::Type::IntOrReal: return DataType::CppType::RealType;
		case DataType::Type::Null:
		case DataType::Type::Integer:
		case DataType::Type::Reference:
		case DataType::Type::List:
		case DataType::Type::AnyMap:
		case DataType::Type::IntMap:
		case DataType::Type::StringMap:
		case DataType::Type::Map:
		case DataType::Type::Grid:
		case DataType::Type::Stack:
		case DataType::Type::Priority: return DataType::CppType::IntType;
		case DataType::Type::Bool: return DataType::CppType::BoolType;
		case DataType::Type::String: return DataType::CppType::StringType;
		case DataType::Type::Vector: return DataType::CppType::VecType;
		case DataType::Type::Matrix: return DataType::CppType::MatrixType;
		case DataType::Type::Array: return DataType::CppType::ArrType;
		case DataType::Type::Unknown:
		case DataType::Type::Variant: return DataType::CppType::VarType;
	}

	return DataType::CppType::VarType;
}

DataType::DataType(Type rawType)
{
	if (rawType != Type::Unknown)
		this->assignments.emplace_back(rawType, "", nullptr);
	updateCppType();
}

DataType::DataType(Type rawType, StringId refId)
{
	if (rawType != Type::Unknown)
		this->assignments.emplace_back(rawType, refId, nullptr);
	updateCppType();
}

DataType::DataType(Type rawType, const DataType* containerType)
{
	if (rawType != Type::Unknown)
		this->assignments.emplace_back(rawType, "", containerType);
	updateCppType();
}

DataType::DataType(StringId name)
{
	String nameStr = String(name);
	List<String> split;
	Type rawType = Type::Unknown;

	if (nameStr.contains("<"))
	{
		split = nameStr.replace(">", "").split('<');
		name = split[0];
	}

	switch (name.id())
	{
		case STR(unknown):
			break;
		case STR(VarType):
		case STR(variant):
			rawType = Type::Variant;
			break;
		case STR(null):
			rawType = Type::Null;
			break;
		case STR(void):
			rawType = Type::Void;
			break;
		case STR(BoolType):
		case STR(bool):
			rawType = Type::Bool;
			break;
		case STR(intorreal):
			rawType = Type::IntOrReal;
			break;
		case STR(IntType):
		case STR(int):
			rawType = Type::Integer;
			break;
		case STR(RealType):
		case STR(real):
			rawType = Type::Real;
			break;
		case STR(StringType):
		case STR(string):
			rawType = Type::String;
			break;
		case STR(reference):
			if (split.size() != 2)
			{
				Console::writeLine("FATAL ERROR: Missing ID for reference.");
				std::exit(1);
			}
			this->assignments.emplace_back(Type::Reference, split[1], nullptr);
			updateCppType();
			return;
		case STR(VecType):
		case STR(vec):
			rawType = Type::Vector;
			break;
		case STR(MatrixType):
		case STR(matrix):
			rawType = Type::Matrix;
			break;
		case STR(ArrType):
		case STR(array):
			rawType = Type::Array;
			break;
		case STR(list):
			rawType = Type::List;
			break;
		case STR(anymap):
			rawType = Type::AnyMap;
			break;
		case STR(intmap):
			rawType = Type::IntMap;
			break;
		case STR(stringmap):
			rawType = Type::StringMap;
			break;
		case STR(map):
			rawType = Type::Map;
			break;
		case STR(grid):
			rawType = Type::Grid;
			break;
		case STR(stack):
			rawType = Type::Stack;
			break;
		case STR(priority):
			rawType = Type::Priority;
			break;
		default:
			if (nameStr == "const VarType&" || nameStr == "VarType&")
			{
				rawType = Type::Variant;
				break;
			}
			Console::writeLine("FATAL ERROR: Unknown data type: {0}", nameStr);
			std::exit(1);
			break;
	}

	if (rawType >= Type::Array) // Container
	{
		DataType containerType;
		if (rawType != Type::Array && isRawTypeArray(rawType)) // vec or matrix
			containerType.reset(Type::Real);

		else if (split.size() > 1)
		{
			String stringEnd = "";
			for (int i = 1; i < static_cast<int>(split.size()); i++)
				stringEnd += split[i] + (i < static_cast<int>(split.size()) - 1 ? "<" : "");
			containerType = DataType(stringEnd);
		}
		else
			containerType.reset(Type::Unknown);

		this->assignments.emplace_back(rawType, "", &containerType);
	}
	else if (rawType != Type::Unknown)
		this->assignments.emplace_back(rawType, "", nullptr);

	updateCppType();
}

void DataType::updateCppType()
{
	// Update C++ type
	bool isSet = false;
	for (const Assignment& ass : this->assignments)
	{
		const CppType assType = cppTypeFor(ass.rawType);

		if (!isSet)
		{
			cppType = assType;
			isSet = true;
		}
		else if (cppType != assType)
		{
			cppType = CppType::VarType;
			return;
		}
	}
}

String DataType::toCpp()
{
	if (DataType::allVarType && !DataType::ignoreAllVarType && cppType != CppType::Void)
		return "VarType";

	switch (cppType)
	{
		case CppType::Void: return "void";
		case CppType::RealType: return "RealType";
		case CppType::IntType: return "IntType";
		case CppType::BoolType: return "BoolType";
		case CppType::StringType: return "StringType";
		case CppType::VecType: return "VecType";
		case CppType::MatrixType: return "MatrixType";
		case CppType::ArrType: return "ArrType";
		case CppType::VarType: return "VarType";
	}
	return "";
}

String DataType::toCppMemberMacro()
{
	switch (cppType)
	{
		case CppType::Void: return "";
		case CppType::RealType: return "Real";
		case CppType::IntType: return "Int";
		case CppType::BoolType: return "Bool";
		case CppType::StringType: return "Str";
		case CppType::VecType: return "Vec";
		case CppType::MatrixType: return "Mat";
		case CppType::ArrType: return "Arr";
		case CppType::VarType: return "Var";
	}
	return "";
}

String DataType::toCppDefaultValue()
{
	switch (cppType)
	{
		case CppType::Void: return "";
		case CppType::RealType: return "0.0";
		case CppType::IntType: return "IntType(0)";
		case CppType::BoolType: return "false";
		case CppType::StringType: return "\"\"";
		case CppType::VecType: return "VecType()";
		case CppType::MatrixType: return "MatrixType()";
		case CppType::ArrType: return "ArrType()";
		case CppType::VarType: return "VarType()";
	}
	return "";
}

String DataType::toCppEnum()
{
	switch (cppType)
	{
		case CppType::Void: return "";
		case CppType::RealType: return "REAL_t";
		case CppType::IntType: return "INTEGER_t";
		case CppType::BoolType: return "BOOLEAN_t";
		case CppType::StringType: return "STRING_t";
		case CppType::VecType: return "VECTOR_t";
		case CppType::MatrixType: return "MATRIX_t";
		case CppType::ArrType: return "ARRAY_t";
		case CppType::VarType: return "VARIANT_t";
	}
	return "";
}

bool DataType::isCppVarType()
{
	return cppType == CppType::VarType;
}

String DataType::getAssignmentsString(String tabs)
{
	String varStr = "";
	for (Assignment& ass : this->assignments)
	{
		varStr += tabs + ass.toString();
		if (ass.func != nullptr)
			varStr += " in " + String(ass.func->name) + ":" + String(ass.line);
		varStr += "\n";
	}
	return varStr;
}

String DataType::toString()
{
	if (this->assignments.size() == 0)
		return "Unknown";

	String str = "";
	for (int i = 0; i < static_cast<int>(this->assignments.size()); i++)
		str += (i > 0 ? ", " : "") + this->assignments[i].toString();

	return str;
}

bool DataType::isUnknown()
{
	return (this->assignments.size() == 0);
}

String DataType::typeName(Type rawType)
{
	switch (rawType)
	{
		case Type::Unknown: return STR(Unknown);
		case Type::Variant: return STR(Variant);
		case Type::Null: return STR(Null);
		case Type::Void: return STR(Void);
		case Type::Bool: return STR(Bool);
		case Type::IntOrReal: return STR(IntOrReal);
		case Type::Integer: return STR(Integer);
		case Type::Real: return STR(Real);
		case Type::String: return STR(String);
		case Type::Reference: return STR(Reference);
		case Type::Array: return STR(Array);
		case Type::Vector: return STR(Vector);
		case Type::Matrix: return STR(Matrix);
		case Type::List: return STR(List);
		case Type::AnyMap: return STR(AnyMap);
		case Type::IntMap: return STR(IntMap);
		case Type::StringMap: return STR(StringMap);
		case Type::Map: return STR(Map);
		case Type::Grid: return STR(Grid);
		case Type::Stack: return STR(Stack);
		case Type::Priority: return STR(Priority);
	}

	return STR(Unknown);
}

const DataType& DataType::scalar(Type rawType)
{
	static const std::array<DataType, static_cast<std::size_t>(Type::Priority) + 1> types = []
	{
		std::array<DataType, static_cast<std::size_t>(Type::Priority) + 1> result;
		for (std::size_t i = 0; i < result.size(); ++i)
			result[i].reset(static_cast<Type>(i));
		return result;
	}();

	return types[static_cast<std::size_t>(rawType)];
}

bool DataType::isReal()
{
	if (this->assignments.size() == 0)
		return false;

	for (Assignment& ass : this->assignments)
		if (isRawTypeReal(ass.rawType))
			return true;

	return false;
}

bool DataType::isContainer()
{
	for (Assignment& ass : this->assignments)
		if (ass.rawType >= Type::Array)
			return true;

	return false;
}

DataType::Assignment* DataType::getFirstAssignment(Type rawType)
{
	for (Assignment& ass : this->assignments)
		if (rawType == Type::Unknown ||
			ass.rawType == rawType ||
			(rawType == Type::AnyMap && isRawTypeMap(ass.rawType)))
			return &ass;

	return nullptr;
}

List<DataType::Assignment*> DataType::getAssignments(Type rawType)
{
	List<Assignment*> result = List<Assignment*>();
	for (Assignment& ass : this->assignments)
		if (rawType == Type::Unknown || ass.rawType == rawType)
			result.add(&ass);

	return result;
}

StringId DataType::getUniqueReferenceId()
{
	StringId refId;
	bool isSet = false;
	for (Assignment& ass : this->assignments)
	{
		if (ass.rawType == Type::Reference)
		{
			if (isSet) // Multiple assignments found containing references, exit "any"
				return STR(any);

			refId = ass.refId;
			isSet = true;
		}
	}

	return refId;
}

DataType::Type DataType::getMapType()
{
	for (Assignment& ass : this->assignments)
		if (isRawTypeMap(ass.rawType))
			return ass.rawType;

	return Type::Unknown;
}

bool DataType::assign(const DataType& inputType, Function* func, int line, int containerLevel)
{
	const int maxContainerLevel = 1;

	if (inputType.assignments.size() == 0) // Unknown input
		return false;

	bool changed = false;
	for (const Assignment& inputAss : inputType.assignments) // Iterate input assignments
	{
		bool addNew = true;
		for (Assignment& ass : this->assignments) // Combine/Add assignments
		{
			Type inputRawType = inputAss.rawType;
			if ((ass.rawTypeReal && inputAss.rawTypeReal) || // Both real
				(ass.rawTypeArray && inputAss.rawTypeArray)) // Both array
			{
				addNew = false;
				if (inputRawType > ass.rawType) // Overwrite if larger
				{
					ass.rawType = inputRawType;
					ass.rawTypeReal = inputAss.rawTypeReal;
					ass.rawTypeArray = inputAss.rawTypeArray;
					ass.rawTypeMap = inputAss.rawTypeMap;
					changed = true;
					ass.func = func;
					ass.line = line;
				}
			}
			else if (inputRawType == Type::Reference && ass.rawType == Type::Reference) // Both reference
			{
				if (ass.refId == 0 && inputAss.refId != 0) // Unknown reference
				{
					ass.refId = inputAss.refId;
					changed = true;
					ass.func = func;
					ass.line = line;
					addNew = false;
				}
				else if (inputAss.refId == ass.refId) // Empty input/Same object
					addNew = false;
			}
			else if (ass.rawTypeMap && inputAss.rawTypeMap && inputRawType != ass.rawType) // Both map
			{
				addNew = false;
				if (inputRawType > ass.rawType) // Overwrite if larger
				{
					if (ass.rawType == Type::IntMap) // StringMap + IntMap -> Map
					{
						ass.rawType = Type::Map;
						ass.rawTypeReal = false;
						ass.rawTypeArray = false;
						ass.rawTypeMap = true;
					}
					else
					{
						ass.rawType = inputRawType;
						ass.rawTypeReal = inputAss.rawTypeReal;
						ass.rawTypeArray = inputAss.rawTypeArray;
						ass.rawTypeMap = inputAss.rawTypeMap;
					}
					changed = true;
					ass.func = func;
					ass.line = line;
				}
			}
			else if (ass.rawType == inputRawType) // Type match
				addNew = false;

			// Assign container
			if ((ass.rawType == Type::Array || ass.rawType >= Type::List) &&
				inputAss.containerStorage != nullptr && containerLevel < maxContainerLevel)
			{
				// Assignment copies share their container until one is mutated.
				// This preserves value semantics without deep-copying every resolver
				// temporary and reset.
				if (ass.containerStorage == nullptr || ass.containerStorage.use_count() != 1)
				{
					ass.containerStorage = ass.containerStorage != nullptr
						? std::make_shared<DataType>(*ass.containerStorage)
						: std::make_shared<DataType>();
				}

				if (ass.containerStorage->assign(*inputAss.containerStorage, func, line, containerLevel + 1))
				{
					changed = true;
					ass.func = func;
					ass.line = line;
				}
			}

			if (!addNew)
				break;
		}

		// Other assignment is unique, add a copy to this type
		if (addNew)
		{
			this->assignments.emplace_back(inputAss, func, line);
			changed = true;
		}
	}

	if (changed)
		updateCppType();

	return changed;
}

void DataType::reset(Type rawType)
{
	this->cppType = cppTypeFor(rawType);
	if (rawType == Type::Unknown)
	{
		this->assignments.clear();
	}
	else if (this->assignments.empty())
	{
		this->assignments.emplace_back(rawType, "", nullptr);
	}
	else
	{
		this->assignments.resize(1);
		Assignment& assignment = this->assignments.front();
		assignment.rawType = rawType;
		assignment.rawTypeReal = DataType::isRawTypeReal(assignment.rawType);
		assignment.rawTypeArray = DataType::isRawTypeArray(assignment.rawType);
		assignment.rawTypeMap = DataType::isRawTypeMap(assignment.rawType);
		assignment.refId.clear();
		assignment.containerStorage.reset();
		assignment.func = nullptr;
		assignment.line = 0;
	}
}

void DataType::reset(Type rawType, const StringId& refId)
{
	this->cppType = cppTypeFor(rawType);
	if (rawType == Type::Unknown)
	{
		this->assignments.clear();
	}
	else if (this->assignments.empty())
	{
		this->assignments.emplace_back(rawType, refId, nullptr);
	}
	else
	{
		this->assignments.resize(1);
		Assignment& assignment = this->assignments.front();
		assignment.rawType = rawType;
		assignment.rawTypeReal = DataType::isRawTypeReal(assignment.rawType);
		assignment.rawTypeArray = DataType::isRawTypeArray(assignment.rawType);
		assignment.rawTypeMap = DataType::isRawTypeMap(assignment.rawType);
		assignment.refId = refId;
		assignment.containerStorage.reset();
		assignment.func = nullptr;
		assignment.line = 0;
	}
}

void DataType::reset(Type rawType, const DataType& containerType)
{
	this->cppType = cppTypeFor(rawType);
	if (rawType == Type::Unknown)
	{
		this->assignments.clear();
	}
	else if (this->assignments.empty())
	{
		this->assignments.emplace_back(rawType, "", &containerType);
	}
	else
	{
		this->assignments.resize(1);
		Assignment& assignment = this->assignments.front();
		assignment.rawType = rawType;
		assignment.rawTypeReal = DataType::isRawTypeReal(assignment.rawType);
		assignment.rawTypeArray = DataType::isRawTypeArray(assignment.rawType);
		assignment.rawTypeMap = DataType::isRawTypeMap(assignment.rawType);
		assignment.refId.clear();
		if (assignment.containerStorage != nullptr && assignment.containerStorage.use_count() == 1)
			assignment.containerStorage->reset(containerType);
		else
			assignment.containerStorage = std::make_shared<DataType>(containerType);
		assignment.func = nullptr;
		assignment.line = 0;
	}
}

void DataType::reset(const DataType& other)
{
	if (this == &other)
		return;

	this->assignments.resize(other.assignments.size());
	for (int i = 0; i < static_cast<int>(other.assignments.size()); i++)
		this->assignments[i] = other.assignments[i];
	this->cppType = other.cppType;
}

DataType::Assignment::Assignment(const DataType::Assignment& other, Function* func, int line)
	: rawType(other.rawType),
	refId(other.refId),
	containerStorage(other.containerStorage),
	func(func),
	line(line),
	rawTypeReal(other.rawTypeReal),
	rawTypeArray(other.rawTypeArray),
	rawTypeMap(other.rawTypeMap)
{
}

DataType::Assignment::Assignment(Type type, StringId refId, const DataType* containerType, Function* func, int line)
	: rawType(type),
	refId(refId),
	containerStorage(containerType ? std::make_shared<DataType>(*containerType) : nullptr),
	func(func),
	line(line)
{
	this->rawTypeReal = DataType::isRawTypeReal(this->rawType);
	this->rawTypeArray = DataType::isRawTypeArray(this->rawType);
	this->rawTypeMap = DataType::isRawTypeMap(this->rawType);
}

String DataType::Assignment::toString()
{
	if (this->rawType >= Type::Array) // Container
		return DataType::typeName(this->rawType) + "<" + this->containerStorage->toString() + ">";
	else if (this->rawType == Type::Reference) // Reference
		return DataType::typeName(this->rawType) + "<" + String(this->refId) + ">";
	else
		return DataType::typeName(this->rawType);
}

}
